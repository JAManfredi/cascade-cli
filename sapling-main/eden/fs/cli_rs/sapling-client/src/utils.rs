/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::HashMap;
use std::env;
use std::ffi::OsString;
use std::fs::read_to_string;
use std::path::Path;

use crate::types::SaplingStatus;

pub(crate) fn get_sapling_executable_path() -> String {
    let path = env::var("EDEN_HG_BINARY").unwrap_or_else(|_| String::new());
    if path.is_empty() {
        "hg".to_string() // `sl` is not always available, so use `hg`
    } else {
        path
    }
}

pub(crate) fn get_sapling_options() -> HashMap<OsString, OsString> {
    let mut options = HashMap::<OsString, OsString>::new();
    // Ensure that the hgrc doesn't mess with the behavior of the commands that we're running.
    options.insert("HGPLAIN".to_string().into(), "1".to_string().into());
    // Ensure that we do not log profiling data for the commands we are
    // running. This is to avoid a significant increase in the rate of logging.
    options.insert("NOSCMLOG".to_string().into(), "1".to_string().into());
    // chg can elect to kill all children if an error occurs in any child.
    // This can cause commands we spawn to fail transiently.  While we'd
    // love to have the lowest latency, the transient failure causes problems
    // with our ability to deliver notifications to our clients in a timely
    // manner, so we disable the use of chg for the sapling processes
    // that we spawn.
    options.insert("CHGDISABLE".to_string().into(), "1".to_string().into());
    options
}

pub fn is_fbsource_checkout(mount_point: &Path) -> bool {
    let project_id_path = mount_point.join(".projectid");
    let project_id = read_to_string(project_id_path).ok();
    match project_id {
        Some(project_id) => project_id.trim() == "fbsource",
        None => false,
    }
}

//
// Single line looks like:
//    <status> <path>
//
// Where status is one of:
//   M = modified
//   A = added
//   R = removed
//   C = clean
//   ! = missing (deleted by a non-sl command, but still tracked)
//   ? = not tracked
//   I = ignored
//     = origin of the previous file (with --copies)
// Note:
//   Paths can have spaces, but are not quoted.
pub(crate) fn process_one_status_line(
    line: &str,
) -> anyhow::Result<Option<(SaplingStatus, String)>> {
    // Must include a status and at least one char path.
    let mut chars = line.chars();
    let status = chars
        .next()
        .ok_or_else(|| anyhow::anyhow!("Invalid status line: {line}"))?;
    let space = chars
        .next()
        .ok_or_else(|| anyhow::anyhow!("Invalid status line: {line}"))?;
    if space != ' ' {
        return Err(anyhow::anyhow!("Invalid status line: {line}"));
    }
    let path = line.chars().skip(1).collect::<String>().trim().to_owned();
    let result = match status {
        'M' => Some((SaplingStatus::Modified, path)),
        'A' => Some((SaplingStatus::Added, path)),
        'R' => Some((SaplingStatus::Removed, path)),
        'C' => Some((SaplingStatus::Clean, path)),
        '!' => Some((SaplingStatus::Missing, path)),
        '?' => Some((SaplingStatus::NotTracked, path)),
        'I' => Some((SaplingStatus::Ignored, path)),
        ' ' => Some((SaplingStatus::Copied, path)),
        _ => None, // Skip all others
    };

    Ok(result)
}

#[cfg(test)]
pub(crate) mod tests {
    use std::io::Error;
    use std::io::ErrorKind;

    use async_process_traits::MockChild;
    use async_process_traits::MockChildHandle;
    use async_process_traits::MockCommandSpawner;
    use async_process_traits::MockExitStatus;
    use edenfs_client::utils::get_mount_point;
    use tokio::io;
    use tokio_test::io::Builder as MockIoBuilder;

    use crate::types::*;
    use crate::utils::*;

    #[test]
    pub fn test_is_fbsource_checkout() -> anyhow::Result<()> {
        let mount_point = get_mount_point(&None)?;
        assert!(is_fbsource_checkout(&mount_point));
        Ok(())
    }

    #[test]
    fn test_process_one_status_line() -> anyhow::Result<()> {
        assert_eq!(
            process_one_status_line("M buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::Modified,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("A buck2/app/buck2_file_watcher/src/edenfs/interface.rs")?,
            Some((
                SaplingStatus::Added,
                "buck2/app/buck2_file_watcher/src/edenfs/interface.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("R buck2/app/buck2_file_watcher/src/edenfs/utils.rs")?,
            Some((
                SaplingStatus::Removed,
                "buck2/app/buck2_file_watcher/src/edenfs/utils.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("! buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::Missing,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("? buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::NotTracked,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("C buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::Clean,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("I buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::Ignored,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        assert_eq!(
            process_one_status_line("  buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")?,
            Some((
                SaplingStatus::Copied,
                "buck2/app/buck2_file_watcher/src/edenfs/sapling.rs".to_owned()
            ))
        );

        // Space in path
        assert_eq!(
             process_one_status_line("M ovrsource-legacy/unity/socialvr/modules/wb_unity_asset_bundles/Assets/MetaHorizonUnityAssetBundle/Editor/Unity Dependencies/ABDataSource.cs")?,
             Some((
                 SaplingStatus::Modified,
                 "ovrsource-legacy/unity/socialvr/modules/wb_unity_asset_bundles/Assets/MetaHorizonUnityAssetBundle/Editor/Unity Dependencies/ABDataSource.cs".to_owned()
             ))
         );

        // Invalid status
        assert!(process_one_status_line("Invalid status").is_err());

        // Invalid status (missing status), but valid path with space in it
        assert!(
             process_one_status_line(" ovrsource-legacy/unity/socialvr/modules/wb_unity_asset_bundles/Assets/MetaHorizonUnityAssetBundle/Editor/Unity Dependencies/ABDataSource.cs")
             .is_err());

        // Malformed status (no space)
        assert!(
            process_one_status_line("Mbuck2/app/buck2_file_watcher/src/edenfs/sapling.rs").is_err()
        );

        // Malformed status (colon instead of space)
        assert!(
            process_one_status_line("M:buck2/app/buck2_file_watcher/src/edenfs/sapling.rs")
                .is_err()
        );

        Ok(())
    }

    pub(crate) fn get_mock_spawner(
        program: String,
        output: Option<(i32, Option<Vec<u8>>)>,
    ) -> MockCommandSpawner {
        MockCommandSpawner::with_callback(move |cmd| match cmd.program.to_str() {
            Some(cmd_program) if cmd_program == program && output.is_some() => {
                let (exit_code, stdout_lines) = output.clone().unwrap();
                Ok(mock_child(exit_code, stdout_lines))
            }
            program => Err(Error::new(
                ErrorKind::Other,
                anyhow::anyhow!("Not expected program: {:?}", program),
            )),
        })
    }

    fn mock_child(exit_code: i32, stdout_lines: Option<Vec<u8>>) -> MockChild {
        let handle = MockChildHandle::new();
        handle.set_status(Ok(Some(MockExitStatus::new(Some(exit_code)))));
        if let Some(stdout_lines) = stdout_lines {
            let stdout = MockIoBuilder::new().read(&stdout_lines).build();
            MockChild::with_stdio(handle, Some(io::sink()), Some(stdout), Some(io::empty()))
        } else {
            MockChild::new(handle)
        }
    }
}
