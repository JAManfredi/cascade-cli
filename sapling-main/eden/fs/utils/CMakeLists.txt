# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

file(GLOB UTILS_SRCS "*.cpp")

if (WIN32)
  list(
    REMOVE_ITEM UTILS_SRCS
    ${CMAKE_CURRENT_SOURCE_DIR}/FutureUnixSocket.cpp
    ${CMAKE_CURRENT_SOURCE_DIR}/IoFuture.cpp
    ${CMAKE_CURRENT_SOURCE_DIR}/SSLContext.cpp
    ${CMAKE_CURRENT_SOURCE_DIR}/Thread.cpp
    ${CMAKE_CURRENT_SOURCE_DIR}/UnixSocket.cpp
    ${CMAKE_CURRENT_SOURCE_DIR}/XAttr.cpp
  )
endif()

add_library(
  eden_utils STATIC
    ${UTILS_SRCS}
)

target_link_libraries(
  eden_utils
  PUBLIC
    eden_build_config
    eden_digest
    eden_service_thrift_cpp
    Folly::folly
    edencommon::edencommon_utils
)

if (WIN32)
  target_link_libraries(
    eden_utils
    PUBLIC
      Userenv.lib
      Rpcrt4.lib
      dbghelp.lib
  )
endif()

add_subdirectory(test)
