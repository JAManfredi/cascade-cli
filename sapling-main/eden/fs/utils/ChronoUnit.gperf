struct ChronoUnit;
%struct-type
%language=C++
%define class-name ChronoUnitPerfectHash
%readonly-tables
%compare-strncmp
%omit-struct-type
%{
/*
 * Copyright (c) Meta Platforms, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */
#include "eden/fs/utils/ChronoUnit.h"

using facebook::eden::ChronoUnit;
using folly::StringPiece;

namespace {
%}
%%
ns, 1, 1000000000
nanosecond, 1, 1000000000
nanoseconds, 1, 1000000000
us, 1, 1000000
# microseconds with a UTF-8 encoded Greek Mu character
"\xce\xbcs", 1, 1000000
microsecond, 1, 1000000
microseconds, 1, 1000000
ms, 1, 1000
millisecond, 1, 1000
milliseconds, 1, 1000
s, 1, 1
second, 1, 1
seconds, 1, 1
m, 60, 1
min, 60, 1
minute, 60, 1
minutes, 60, 1
h, 3600, 1
hr, 3600, 1
hour, 3600, 1
hours, 3600, 1
d, 86400, 1
day, 86400, 1
days, 86400, 1
wk, 604800, 1
week, 604800, 1
weeks, 604800, 1
# C++20 defines std::chrono::months as exactly 1/12 of 1 year
mon, 2629746, 1
month, 2629746, 1
months, 2629746, 1
# C++20 defines std::chrono::years as exactly 365.2425 days
yr, 31556952, 1
year, 31556952, 1
years, 31556952, 1
%%
} // unnamed namespace

namespace facebook {
namespace eden {

const ChronoUnit* FOLLY_NULLABLE lookupChronoUnitInfo(StringPiece unitName) {
  return ChronoUnitPerfectHash::in_word_set(unitName);
}

} // namespace eden
} // namespace facebook
