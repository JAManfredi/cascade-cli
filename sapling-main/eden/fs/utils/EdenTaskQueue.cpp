/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#include "eden/fs/utils/EdenTaskQueue.h"

namespace facebook::eden {

folly::BlockingQueueAddResult EdenTaskQueue::add(
    folly::CPUThreadPoolExecutor::CPUTask item) {
  queue_.enqueue(std::move(item));
  return sem_.post();
}

folly::CPUThreadPoolExecutor::CPUTask EdenTaskQueue::take() {
  sem_.wait();
  folly::CPUThreadPoolExecutor::CPUTask res;
  queue_.dequeue(res);
  return res;
}

folly::Optional<folly::CPUThreadPoolExecutor::CPUTask>
EdenTaskQueue::try_take_for(std::chrono::milliseconds time) {
  if (!sem_.try_wait_for(time)) {
    return folly::none;
  }
  folly::CPUThreadPoolExecutor::CPUTask res;
  queue_.dequeue(res);
  return res;
}

size_t EdenTaskQueue::size() {
  return queue_.size();
}

} // namespace facebook::eden
