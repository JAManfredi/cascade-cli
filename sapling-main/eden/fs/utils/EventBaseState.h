/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#pragma once

#include <folly/io/async/EventBase.h>

namespace facebook::eden {

/**
 * EventBase state machines need to ensure their state is only accessed from the
 * EventBase. EventBaseState provides that guarantee: the state can only be
 * accessed from the given EventBase.
 */
template <typename State>
class EventBaseState {
 public:
  /**
   * Constructs an EventBaseState tied to the specified EventBase.
   */
  template <typename... T>
  explicit EventBaseState(folly::EventBase* evb, T&&... args)
      : evb_{evb}, state_{std::forward<T>(args)...} {}

  EventBaseState(const EventBaseState&) = delete;
  EventBaseState(EventBaseState&&) = delete;
  EventBaseState& operator=(const EventBaseState&) = delete;
  EventBaseState& operator=(EventBaseState&&) = delete;

  State& get() {
    evb_->checkIsInEventBaseThread();
    return state_;
  }

  const State& get() const {
    evb_->checkIsInEventBaseThread();
    return state_;
  }

 private:
  folly::EventBase* evb_;
  State state_;
};

} // namespace facebook::eden
