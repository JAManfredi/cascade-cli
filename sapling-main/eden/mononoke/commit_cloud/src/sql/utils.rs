/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

pub fn prepare_prefix(input: &str) -> String {
    let mut escaped = input.to_string();
    escaped = escaped.replace('\\', "\\\\");
    escaped = escaped.replace('_', "\\_");
    escaped = escaped.replace('%', "\\%");
    escaped.push('%');
    escaped
}
