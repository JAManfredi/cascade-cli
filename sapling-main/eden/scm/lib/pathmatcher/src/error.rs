/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#[derive(Debug, thiserror::Error)]
pub enum Error {
    UnsupportedPatternKind(String),
    PathOutsideRoot(String, String, bool),
    NonUtf8(String),
    StdinUnavailable,
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::UnsupportedPatternKind(s) => write!(f, "unsupported pattern kind {}", s),
            Error::PathOutsideRoot(path, root, show_hint) => {
                let message = format!("cwd relative path '{}' is not under root '{}'", path, root);
                if *show_hint {
                    let hint_message = "consider using --cwd to change working directory";
                    write!(f, "{}\n(hint: {})", message, hint_message)
                } else {
                    write!(f, "{}", message)
                }
            }
            Error::NonUtf8(s) => write!(f, "non-utf8 path '{}' when building pattern", s),
            Error::StdinUnavailable => write!(
                f,
                "listfile:- may only be used once as a direct CLI argument"
            ),
        }
    }
}
