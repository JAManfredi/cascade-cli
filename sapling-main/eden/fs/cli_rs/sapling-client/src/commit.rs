/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use async_process_traits::Command;
use async_process_traits::CommandSpawner;
use async_process_traits::ExitStatus;
use async_process_traits::Output;
use async_process_traits::TokioCommandSpawner;

use crate::error::Result;
use crate::utils::get_sapling_executable_path;
use crate::utils::get_sapling_options;

pub async fn get_current_commit_id() -> Result<String> {
    get_current_commit_id_impl(&TokioCommandSpawner).await
}

async fn get_current_commit_id_impl<Spawner>(spawner: &Spawner) -> Result<String>
where
    Spawner: CommandSpawner,
{
    let mut command = Spawner::Command::new(get_sapling_executable_path());
    command
        .envs(get_sapling_options())
        .args(["whereami", "--traceback"]);
    let output = spawner.output(&mut command).await?;

    Ok(String::from_utf8(output.stdout().to_vec())?)
}

pub async fn get_commit_timestamp(commit_id: &str) -> Result<u64> {
    get_commit_timestamp_impl(&TokioCommandSpawner, commit_id).await
}

async fn get_commit_timestamp_impl<Spawner>(spawner: &Spawner, commit_id: &str) -> Result<u64>
where
    Spawner: CommandSpawner,
{
    let mut command = Spawner::Command::new(get_sapling_executable_path());
    command.envs(get_sapling_options()).args([
        "log",
        "--traceback",
        "-T",
        "{date}",
        "-r",
        commit_id,
    ]);

    let output = spawner.output(&mut command).await?;
    let date = String::from_utf8(output.stdout().to_vec())?;
    let date = date.parse::<f64>()?;
    // NOTE: Sapling returns fractional seconds, but we only want whole seconds.
    //       Truncate to the nearest second.
    Ok(date as u64)
}

pub async fn is_commit_in_repo(commit_id: &str) -> Result<bool> {
    is_commit_in_repo_impl(&TokioCommandSpawner, commit_id).await
}

async fn is_commit_in_repo_impl<Spawner>(spawner: &Spawner, commit_id: &str) -> Result<bool>
where
    Spawner: CommandSpawner,
{
    let mut command = Spawner::Command::new(get_sapling_executable_path());
    command.envs(get_sapling_options()).args([
        "log",
        "--traceback",
        "-r",
        commit_id,
        // Disable fbsource <-> www sync during lookup
        "--config",
        "megarepo.transparent-lookup=",
    ]);

    let output = spawner.output(&mut command).await?;
    Ok(output.status().success())
}

#[cfg(test)]
mod tests {
    use crate::commit::*;
    use crate::utils::tests::get_mock_spawner;

    const COMMIT_ID: &str = "0000111122223333444455556666777788889999";

    #[tokio::test]
    pub async fn test_get_current_commit_id() -> Result<()> {
        let spawner = get_mock_spawner(
            get_sapling_executable_path(),
            Some((0, Some(COMMIT_ID.as_bytes().to_vec()))),
        );
        let commit_id = get_current_commit_id_impl(&spawner).await?;
        assert_eq!(commit_id, COMMIT_ID);
        Ok(())
    }

    #[tokio::test]
    pub async fn test_is_commit_in_repo() -> Result<()> {
        let spawner = get_mock_spawner(get_sapling_executable_path(), Some((0, None)));
        assert!(is_commit_in_repo_impl(&spawner, COMMIT_ID).await?);

        let spawner = get_mock_spawner(get_sapling_executable_path(), Some((255, None)));
        assert!(!is_commit_in_repo_impl(&spawner, COMMIT_ID).await?,);

        Ok(())
    }

    #[tokio::test]
    pub async fn test_get_commit_timestamp() -> Result<()> {
        // sl log of commit in fbsource:
        //   changeset:   5496dd87e5fe7430a1a399530cc339a479097524  D68746950
        //   user:        John Elliott <jdelliot@fb.com>
        //   date:        Tue, 28 Jan 2025 10:35:17 -0800
        //   summary:     [meerkat] Improve saved state support
        // timestamp should be 1738089317.028800, but we truncate to the nearest second
        let timestamp_str = "1738089317.028800".to_owned();
        let spawner = get_mock_spawner(
            get_sapling_executable_path(),
            Some((0, Some(timestamp_str.as_bytes().to_vec()))),
        );
        let timestamp = get_commit_timestamp_impl(&spawner, COMMIT_ID).await?;
        assert_eq!(timestamp, 1738089317);
        Ok(())
    }
}
