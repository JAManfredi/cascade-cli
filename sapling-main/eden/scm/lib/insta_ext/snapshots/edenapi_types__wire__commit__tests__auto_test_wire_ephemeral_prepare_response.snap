---
source: commit.rs
expression: "WireEphemeralPrepareResponse::arbitrary(&mut g)"

---
{"1":2939647976913485485}
